bson/json_util.py

```diff
@@ -855,100 +855,189 @@ def _encode_int64(obj: Int64, json_options: JSONOptions) -> Any:
-        except KeyError:
-            raise TypeError("%r is not JSON serializable" % obj) from None
-    elif isinstance(obj, int):
-        if json_options.json_mode == JSONMode.CANONICAL:
-            if -(2**31) <= obj < 2**31:
-                return {"$numberInt": str(obj)}
-            return {"$numberLong": str(obj)}
-        return obj
-    elif isinstance(obj, float):
-        if json_options.json_mode != JSONMode.LEGACY:
-            if math.isnan(obj):
-                return {"$numberDouble": "NaN"}
-            elif math.isinf(obj):
-                representation = "Infinity" if obj > 0 else "-Infinity"
-                return {"$numberDouble": representation}
-            elif json_options.json_mode == JSONMode.CANONICAL:
-                # repr() will return the shortest string guaranteed to produce the
-                # original value, when float() is called on it.
-                return {"$numberDouble": str(repr(obj))}
-        return obj
-    elif isinstance(obj, str):
-        return obj
-    elif isinstance(obj, datetime.datetime):
-        if json_options.datetime_representation == DatetimeRepresentation.ISO8601:
-            if not obj.tzinfo:
-                obj = obj.replace(tzinfo=utc)
-                assert obj.tzinfo is not None
-            if obj >= EPOCH_AWARE:
-                off = obj.tzinfo.utcoffset(obj)
-                if (off.days, off.seconds, off.microseconds) == (0, 0, 0):  # type: ignore
-                    tz_string = "Z"
-                else:
-                    tz_string = obj.strftime("%z")
-                millis = int(obj.microsecond / 1000)
-                fracsecs = ".%03d" % (millis,) if millis else ""
-                return {
-                    "$date": "{}{}{}".format(obj.strftime("%Y-%m-%dT%H:%M:%S"), fracsecs, tz_string)
-                }
-
-        millis = _datetime_to_millis(obj)
-        if json_options.datetime_representation == DatetimeRepresentation.LEGACY:
-            return {"$date": millis}
-        return {"$date": {"$numberLong": str(millis)}}
-    elif isinstance(obj, bytes):
-        return _encode_binary(obj, 0, json_options)
-    elif isinstance(obj, uuid.UUID):
-        if json_options.strict_uuid:
-            binval = Binary.from_uuid(obj, uuid_representation=json_options.uuid_representation)
-            return _encode_binary(binval, binval.subtype, json_options)
-        else:
-            return {"$uuid": obj.hex}
+    # First see if the type is already cached. KeyError will only ever
+    # happen once per subtype.
+    obj_type = type(obj)
+    try:
+        return _ENCODERS[obj_type](obj, json_options)
+    except KeyError:
+        pass
+
+    # Second, fall back to trying _type_marker. This has to be done
+    # before the loop below since users could subclass one of our
+    # custom types that subclasses a python built-in (e.g. Binary)
+    if hasattr(obj, "_type_marker"):
```

Comment:
Is this check explicitly for subclassed BSON types that users define? All of our own BSON types should already be cached in `_ENCODERS`, correct?

Reply:
Correct, I'll add some test cases for this (if they don't already exist).

Reply:
Wow there was no tests for this in BSON, I added one. The use-case is a little convoluted. Not sure if this is something we'll support but at least we have more consistency between bson and json encoding now.

Reply:
Makes sense, I think the comment would be clearer if it explicitly said why this is needed.

Reply:
Oh this comment it clear to me now. It's saying we need to check type_marker first because otherwise a class like this might be encoded as a `bytes` instead of a `Binary` because both `isinstance(MyBin, Binary)` and `isinstance(MyBin, bytes)` would be true and we don't control which check happens first:
```python
class MyBin(Binary):  # Will this be encoded as bytes or as Binary?
    pass
```

I'll keep the comments as they are (to mirror the bson comments) and add this case to the tests.